Add process applications feature

philippfromme opened the pull request:
# Proposed Changes

## Backend

* adds `file-context` feature
  * all open files are added to file context
  * metadata is extracted from every file with a known file extension (`.bpmn`, `.dmn`, `.form`, `.process-application`)
  * error messages are added to files that have a known file extension but either couldn't be read or processed
  * process application is opened/closed automatically if opened/closed file is part of process application
  * opening process application watches process application directory

## Frontend

* `ProcessApplications` feature to open and close process applications automatically
* process applications is a plugin

### ✨ New application menu entry

* to create process application
* choose directory, empty `.process-application` file will be created

<img src="https://github.com/user-attachments/assets/61e2469b-9237-4601-bc16-ffe3f2123295" width="100%">

### ✨ New button in status bar

* shown when open file is part of process application

<img src="https://github.com/user-attachments/assets/fc4b3c99-8711-4f34-9360-dc77dc5bb598" width="100%">

### ✨ New overlay

* shows process application file and reveals it in file explorer on click
* shows all files that are part of same process application as open file and opens them on click
* shows all files that do not have known file extension (`.bpmn`, `.dmn`, `.form`) but have been opened in a tab
* shows all files that have known file extension but have an error (no open on click)

<img width="100%" alt="electron_KY1s2lzIde" src="https://github.com/user-attachments/assets/883c172f-328c-413b-9a2d-85bd5e5ec283" />

### ✨ Tab groups

* active tab now has darker background instead of bottom border
* bottom border for files that are part of process application
* different border colors for different process applications

<img src="https://github.com/user-attachments/assets/396a69ef-4edb-4a51-939c-7c2d3253f4b9" width="100%">

# Try it out

I've added an example process application: https://github.com/camunda/camunda-modeler/pull/4762/commits/a354d21e5cb5da9cf0db8320aa7418650712d3cd

Open any file in this directory to see the process application.

To run the Camunda Modeler try `npm install && npm start`

# Follow-ups

* `References` feature to get reference from and to resources
* `AutoCompletions` feature to get autocompletions
* Branch: https://github.com/camunda/camunda-modeler/tree/references-autocompletions

Closes https://github.com/camunda/camunda-modeler/issues/4666
Closes https://github.com/camunda/camunda-modeler/issues/4675
Closes https://github.com/camunda/camunda-modeler/issues/4687
Related to https://github.com/camunda/camunda-modeler/issues/4676
Related to https://github.com/camunda/camunda-modeler/issues/4664

# Checklist

To ensure you provided everything we need to look at your PR:

* [x] __Brief textual description__ of the changes present
* [x] __Visual demo__ attached
* [x] __Steps to try out__ present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [x] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--
Thanks for creating this pull request! ❤️
-->

nikku commented on client/src/app/App.js:
```diff
@@ -1136,8 +1134,7 @@ export class App extends PureComponent {
       activeTab !== prevState.activeTab ||
       tabs !== prevState.tabs ||
-      layout !== prevState.layout ||
-      endpoints !== prevState.endpoints
```
This will be used (again) once we support "endpoints" for deployment. 

Why would we remove it here?

nikku replied:
(Of course fine, if we consider to store the endpoints in a different place, once supported).

philippfromme replied:
I removed it because it's not used at all at the moment and was probably meant to be removed. I would rather reintroduce it again in the context of adding a feature that requires it.

nikku replied:
We just have to be aware that this is "the data model" that was previously used. We can remove it, but then we'll not remember the data model once we re-introduce an endpoints feature.

philippfromme commented:
Things that aren't working at the moment:

* updates of the process application file itself aren't picked up
*

nikku commented:
@philippfromme Please add "steps to try out" to this PR, so @lmbateman and others can give it a run.

jarekdanielak commented on client/src/app/primitives/Tabbed.less:
```diff
@@ -205,13 +206,17 @@
     }
 
-    &:not(.tab--active):hover::after {
+    &:not(.tab--active):hover:after {
```
This change causes tab close icon to look odd like in my comment: https://github.com/camunda/camunda-modeler/pull/4762#issuecomment-2592277619

philippfromme commented:
> Perhaps if we can now create and open an application process, we could use an option to close it? Similar to close active/all/other tabs:

You mean _Close process application_ would close all files belonging to the open process application? Interesting idea I haven't thought about. 🤔

jarekdanielak commented:
When I try to deploy a file from a process application, the modeler just reloads. 

https://github.com/user-attachments/assets/02217d81-dfb8-4468-a89e-ed615d76dd2d

philippfromme commented:
> When I try to deploy a file from a process application, the modeler just reloads.
> 
> Same thing happens when I save a file. Both only happens to files in a process application, the rest works normally.
> 
>  Screen.Recording.2025-01-15.at.11.56.30.mov

This happens becasue the modeler reloads when something in `resources` changes. 🙈 Opening a process application outside will prevent this.

barmac commented:
I've added the `bpmn/dmn-moddle` libraries to the `app` so that a built application can run: https://github.com/camunda/camunda-modeler/pull/4762/commits/70f0555dc2e7fcccbcab7be0907dd19af85ceb1a

barmac commented:
I'll try to help with this one.

barmac commented:
@philippfromme were you able to run a local build of this? So the one from `npm run build`.

On my machine, the build job results in the source code of `dmn-moddle` being included in the asar. This is not coming from `node_modules` where the package is installed correctly.

To verify the build, besides running the app, you can execute this command for MacOS:
`npx asar list dist/mac-arm64/Camunda\ Modeler.app/Contents/Resources/app.asar | grep dmn-moddle`
This should show `dist` in the list. Not sure how to do it on Windows.

jarekdanielak commented:
Creating a variable on any task in BPMN model with Variables panel open (or opening it after) crashes.

https://github.com/user-attachments/assets/500d6e03-9ae1-421c-9bba-71123f925bd5

jarekdanielak commented:
@barmac I was able to successfully build locally and run the app. Hit me up if you want to test stuff.

barmac commented:
OK so it seems to be my local configuration problem. It's not a blocker in that case.

philippfromme commented:
Converted back to draft. Based on the discussion with @barmac I will make some changes to the implementation:

* try to make process applications (or at least its UI) a plugin
* add the ability for plugins to register actions instead of having to hard-code them in `App.js`
* try not to pollute the global state in `App.js` with state only relevant for process applications

![image](https://github.com/user-attachments/assets/3f888299-fa98-4df8-b91b-2fd4ecc424e5)

Miro: https://miro.com/app/board/uXjVPmTBc04=/?share_link_id=716131412171

barmac commented:
@philippfromme Let's look into this today. 

I checked out the UI today, and I think having "New process application" on top of `Cmd+N` menu is suboptimal, because this is the action that I'd rather not expect to happen that often. We create a process application once, and then add multiple diagrams.

lmbateman commented:
I haven't had time to do a full review, but here are some initial thoughts/findings:

- When I select **File > New Process Application** and then select a folder, there’s no confirmation that my process application has been created. I wasn’t sure if it had been, so I tried **File > New Process Application** again and tried to select a file inside the folder, but they were all grayed out. I wasn’t sure if my process application had been created until I opened one of the files and saw the process application icon in the bottom left corner. I found this process a little confusing.
- Would it be helpful to have an Open All command in the popup that lists the process application files? (But I see it stays open even after I click on a filename, so maybe we don’t need Open All.)
- It might be helpful to have the folder name as the header in the popup, instead of “Process Application”.
- I made my Downloads folder a process application, and apparently the popup doesn’t scroll. :D

![image](https://github.com/user-attachments/assets/eb1642de-889c-46f1-8689-ab9a688f9d9d)

philippfromme commented:
@lmbateman Good feedback. I agree, creating a new process application leaves you with no indication whether it worked or not. We could show a success notification. When creating a new process application only directories can be selected, not files. We could make this more forgiving, allow files as well and then use the parent directory of the file (select foo/bar/baz.bpmn - foo/bar/.process-application will be created). Showing the directory name in the popup is something I'll try.

lmbateman commented:
@philippfromme I think a success notification would work, maybe with the process application name (e.g., "Successfully created process application <PA Name> with N files." I like the idea of being able to open a PA by selecting a file (if one exists) -- I think it will be rare for users to open a process application and not open a file to work on -- but would be interested to hear feedback from the rest of the team.

I'm hoping to do a more thorough review this week.

philippfromme commented:
I've fixed the scrolling issue and added the process application path to the overlay. Will update the pull request shortly.

![image](https://github.com/user-attachments/assets/242a31ec-4af0-48f8-8f13-825bd8135cbf)

philippfromme commented:
I tried out grouping by bottom border color and using a darker background for the selected tab.

![image](https://github.com/user-attachments/assets/0b2fbe4a-4469-479e-b426-ec0eabf1b960)

philippfromme commented:
Ready for another review.

barmac commented:
I've just found a way to crash the Modeler with PA, namely via non-standard file extension (BPMN with `.txt`):

![image](https://github.com/user-attachments/assets/ed74fa97-b472-4ad8-87d7-a85e093bf87b)

```
INFO app:file-context added [
  '/Users/maciej/Downloads/test-diagram.bpmn.txt',
  '/Users/maciej/Desktop/.process-application'
]
INFO app:file-context removed []
INFO app:file-context items [
  {
    uri: 'file:///Users/maciej/Downloads/test-diagram.bpmn.txt',
    metadata: undefined
  },
  {
    uri: 'file:///Users/maciej/Desktop/.process-application',
    metadata: '{\n  "type": "processApplication"\n}'
  }
]
```

barmac commented:
I see two problems to solve:
* the front-end should not crash for the formats that it does not recognize
* the back-end should attach proper meta-data regardless of the file extension

barmac commented:
As a next step, we should let the processors decide whether they can handle the file: `processor.accept(file)` or

```js
if (processor.canHandle(file)) { processor.process(file) }
```

nikku commented:
CI is failing, code looks good enough for the first iteration.

nikku commented:
@barmac Consider to add a `bpmn20.bpmn.txt` file to the example (https://github.com/camunda/camunda-modeler/pull/4762/commits/e08e695b02b8964f7a8a9cf13fec11adf5d216be).

barmac commented:
> @barmac Consider to add a `bpmn20.bpmn.txt` file to the example ([e08e695](https://github.com/camunda/camunda-modeler/commit/e08e695b02b8964f7a8a9cf13fec11adf5d216be)).

Will do.

barmac commented:
One thing which slipped is that we link to [the Web Modeler documentation](https://docs.camunda.io/docs/components/modeler/web-modeler/process-applications/). I think we should have a dedicated page for Desktop Modeler implementation. Perhaps a separate page on the concept itself would make sense once the feature is available for both Modelers.

barmac commented:
I'd rather not make it the top option as it is unlikely to create multiple PAs in one session.

![image](https://github.com/user-attachments/assets/de52f605-df70-4f34-9227-b245b72fafc1)

barmac commented:
I've discovered a way to crash it with looping symlinks:

![image](https://github.com/user-attachments/assets/35d6dda3-03a8-4e08-a948-8ab9da0546b6)

```
❯ tree
.
├── diagram_1.bpmn
├── test
│   ├── form.json
│   ├── test.bpmn
│   └── test2 -> ../test2
└── test2
    └── test -> ../test
```

`test.bpmn` was an empty file, this led to errors:

```
  {
    uri: 'file:///private/tmp/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test.bpmn',
    metadata: '{\n' +
      '  "type": "bpmn",\n' +
      '  "error": "unparsable content detected\\n\\tline: 0\\n\\tcolumn: 0\\n\\tnested error: missing start tag",\n' +
      '  "ids": [],\n' +
      '  "linkedIds": []\n' +
      '}'
  },
  {
    uri: 'file:///private/tmp/test/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test.bpmn',
    metadata: '{\n' +
      '  "type": "bpmn",\n' +
      '  "error": "unparsable content detected\\n\\tline: 0\\n\\tcolumn: 0\\n\\tnested error: missing start tag",\n' +
      '  "ids": [],\n' +
      '  "linkedIds": []\n' +
      '}'
  }
]
INFO app:file-context workqueue:empty
(node:72618) UnhandledPromiseRejectionWarning: Error: ELOOP: too many symbolic links encountered, stat '/private/tmp/test/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2'
(node:72618) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 37)
(node:72618) UnhandledPromiseRejectionWarning: Error: ELOOP: too many symbolic links encountered, stat '/private/tmp/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test/test2/test'
(node:72618) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 38)
```

We should disable `followSymlinks` option on file watcher.

barmac commented:
Also, we assume that a form json contains `id` property. We should try-catch this as `null` is a valid JSON.

barmac commented:
^ - these are small changes

barmac commented:
Tests pending...

barmac commented:
Symlink issue can only be partially resolved with configuration. Even with `followSymlink` we report a file twice.

barmac commented:
> Symlink issue can only be partially resolved with configuration. Even with `followSymlink` we report a file twice. There's no loop though.

OK somehow it's good now o.O

philippfromme commented:
> Also, we assume that a form json contains `id` property. We should try-catch this as `null` is a valid JSON.

A form always has an ID by default and you cannot remove it through the properties panel. If we want to account for cases that cannot be modeled we should also account for processes and decisions with no ID and add the same error handling to the BPMN and DMN processor. 🤔

philippfromme commented:
With the new RPA tab merged in the meantime should we add these as recognized process application items?

jarekdanielak commented:
> With the new RPA tab merged in the meantime should we add these as recognized process application items?

Perhaps only if RPA is enabled with the flag?

barmac commented:
> With the new RPA tab merged in the meantime should we add these as recognized process application items?

Eventually, yes. But it's also OK to add the support later, when we understand the story of RPA linking. AFAIK it was done via element templates so far which indicates that we may be missing something (@marstamm correct me if I'm wrong).

philippfromme commented:
> OK I've used the warning icon. We could do better, but at least it's not failing.

We could get rid of the need for extensions in the backend by reusing the `TabsProvider`.

### Current

![Camunda Modeler Process Applications - Old](https://github.com/user-attachments/assets/eaca0dc7-1cb6-4336-9707-a88afac732cc)


### Potential solution

![Camunda Modeler Process Applications - New](https://github.com/user-attachments/assets/c80abb76-b5d2-4503-81b7-3b95193141a3)

nikku commented:
> This is definitely not the behavior we want. If a file can be opened just fine why would it show up with a warning in the process application files.

:+1:

barmac commented:
> A form always has an ID by default and you cannot remove it through the properties panel. If we want to account for cases that cannot be modeled we should also account for processes and decisions with no ID and add the same error handling to the BPMN and DMN processor. 🤔

I'd like us to not crash even on malformed files. Note that even for `.json` files these can be forms as well as `package.json`.

barmac commented:
> This is definitely not the behavior we want. If a file can be opened just fine why would it show up with a warning in the process application files.

I agree. This was only a mitigation for the extensions-based mechanism. I think we want to eventually use the same logic for `canOpen` in the frontend and in file context.

philippfromme commented:
I've [tried to fix](https://github.com/camunda/camunda-modeler/pull/4762/commits/e40cf7df648b5e59681679f9a982efb155f0aceb) https://github.com/camunda/camunda-modeler/pull/4762#issuecomment-2685124711 but ran into some issues:

* what files can be opened is ultimately decided by the [TabsProvider](https://github.com/camunda/camunda-modeler/blob/develop/client/src/app/TabsProvider.js#L646), if it doesn't create a tab for a file, the file cannot be opened
* [plugins can be used](https://github.com/camunda/camunda-modeler/blob/develop/client/src/app/TabsProvider.js#L144) to open other files (we'll ignore this for process applications)
* the TabsProvider is frontend only, it's not reusable in the backend
* I tried to work around this problem for now by essentially duplicating (🙈) the [`canOpen`](https://github.com/camunda/camunda-modeler/blob/develop/client/src/app/TabsProvider.js#L170) functions just to get it to work (we currently have no concept of using things both in the frontend _and_ backend, we use two different module systems)
* so far we [ignored all files with an unknown extension](https://github.com/camunda/camunda-modeler/pull/4762/commits/f420212fa6e9238588fa6695a0ddc0c78417a145#diff-c69e76babb28b9a37e44314288983f593583eb1f34bdedcd7f62660ea3390fc3R60) meaning we didn't read and process them at all (good for performance)
* TabsProvider decides whether a file can be opened not only based on the extension but also based on the content which requires reading and processing the file
* if [we adopt this strategy](https://github.com/camunda/camunda-modeler/pull/4762/commits/e40cf7df648b5e59681679f9a982efb155f0aceb#diff-1e0b085121db53bd7c329857870e29a2565069b95323e998281db888e145fd84R280) this will have a huge impact on performance
* I've tried opening a file in a directory with a lot of different files and so far it just crashes the application, I haven't been able to determine whether it's a bug or due to the performance impact

barmac commented:
Regarding performance, perhaps a good compromise could be to still filter the files by extension, but index the file which was correctly opened on the frontend? The latter part would fail previously because the application tested the extension to decide on the processor.

barmac commented:
We could also allow to provide the file type to the indexer, sth like `fileContext.add(file, type?)`

barmac commented on client/src/app/App.js:
```diff
@@ -717,6 +743,8 @@ export class App extends PureComponent {
 
   readFileFromPath = async (filePath) => {
 
+    debugger;
```
Note that linter has already caught it: https://github.com/camunda/camunda-modeler/actions/runs/13584133119/job/37975368629?pr=4762#step:6:25

philippfromme commented:
I've made adjustments based on our discussion and the [suggested solution](https://github.com/camunda/camunda-modeler/pull/4762#issuecomment-2690120075).

* files are indexed (automatically) only if they have a known file extension
* files are indexed explicitly if they have been opened

Ready for review.

jarekdanielak commented on app/lib/menu/menu-builder.js:
```diff
@@ -268,6 +269,17 @@ class MenuBuilder {
 
+  appendNewProcessApplication() {
+    this.menu.append(new MenuItem({
+      label: 'New Process Application',
```
```suggestion
      label: 'New Process Application...',
```
The dots suggest that it opens a new window.

jarekdanielak commented:
Small UX things: 

The path preview is not really helpful, cause I can't copy it or open the containing directory. I can reveal the files in the file explorer from the tabs bar, so perhaps here it would make sense to just show the name of the directory where `.process-application` file sits which is de facto the name of the process application?

![image](https://github.com/user-attachments/assets/68464852-8eee-4941-b980-ce3ea647af3a)

When I'm creating a new process application on macOS, I have to have the target directory already created, cause there is no option to create it from the pop up. Compare it with the second screenshot where I can create directories on the go (Save menu from Chrome). I don't know how complicated this is in Electron though.

![image](https://github.com/user-attachments/assets/78e04871-c82e-4b09-ac30-4038b92af1d1)

![image](https://github.com/user-attachments/assets/bf4c1280-97aa-46fd-be9d-ece7d2f02ea7)

philippfromme commented:
> The path preview is not really helpful, cause I can't copy it or open the containing directory. I can reveal the files in the file explorer from the tabs bar, so perhaps here it would make sense to just show the name of the directory where .process-application file sits which is de facto the name of the process application?

Maybe we could open the director on click?

philippfromme commented:
@jarekdanielak I tried a different approach:

* show file icon and name instead of full path
* make it clickable like other files in overlay
* on click reveal in file explorer instead of opening tab

![electron_L286UuzHYd](https://github.com/user-attachments/assets/64656745-6a37-4a20-acf6-73313b109483)

barmac commented:
@philippfromme to create an issue for RPA integration.

philippfromme commented:
Follow-up issues:

* https://github.com/camunda/camunda-modeler/issues/4892
* https://github.com/camunda/camunda-modeler/issues/4893

barmac commented:
I found a way to break it:

![image](https://github.com/user-attachments/assets/53d4e2e6-dd27-47ef-899d-144c2111be34)

https://github.com/user-attachments/assets/0b3b7bd9-ece3-4d08-9357-f79da49ac5a0

barmac commented:
BTW this is not related to symlink

barmac commented:
So to reproduce the problem, you need to open an empty file.

philippfromme commented:
@barmac I fixed the issue you reported. Error handling has been improved so empty files now aren't shown as an error and can be opened.

<img width="441" alt="electron_KY1s2lzIde" src="https://github.com/user-attachments/assets/605e0b35-2b49-4512-a9cb-0d2736ef3e57" />

I also increased the test coverage so Codecov is happy.

barmac reviewed (approved):
Let's get this merged!

barmac commented:
To add: "Closes..." to the commit closing issues.
